File: creator-node/src/monitors/filesystem.js
const si = require('systeminformation')
const disk = require('diskusage')
const DiskManager = require('../diskManager')

const getStoragePathSize = async () => {
  const storagePath = DiskManager.getConfigStoragePath()
  const { total } = await disk.check(storagePath)
  return total
}

const getStoragePathUsed = async () => {
  const storagePath = DiskManager.getConfigStoragePath()
  const { available, total } = await disk.check(storagePath)
  return total - available
}

// We first check '/var/k8s' in case the service operator has elected to
// mount an external volume for k8s data. Otherwise, default to the root path at '/'

const getFilesystemSize = async () => {
  const fsSizes = await si.fsSize()

  let fsSize = fsSizes.find((fsSize) => fsSize.mount === '/var/k8s')
  if (!fsSize) {
    fsSize = fsSizes.find((fsSize) => fsSize.mount === '/')
  }

  return fsSize.size
}

const getFilesystemUsed = async () => {
  const fsSizes = await si.fsSize()

  let fsSize = fsSizes.find((fsSize) => fsSize.mount === '/var/k8s')
  if (!fsSize) {
    fsSize = fsSizes.find((fsSize) => fsSize.mount === '/')
  }

  return fsSize.used
}

// TODO: Determine how to compute IOPS -- `systeminformation`
// doesn't handle this gracefully when containerized.
//
// const getFilesystemReadsPerSec = async () => {
//   const fsStats = await si.fsStats()
//   return fsStats.rx_sec
// }
// const getFilesystemWritesPerSec = async () => {
//   const fsStats = await si.fsStats()
//   return fsStats.wx_sec
// }

const getMaxFileDescriptors = async () => {
  const fsOpenFiles = await si.fsOpenFiles()
  return fsOpenFiles.max
}

const getAllocatedFileDescriptors = async () => {
  const fsOpenFiles = await si.fsOpenFiles()
  return fsOpenFiles.allocated
}

module.exports = {
  getStoragePathSize,
  getStoragePathUsed,
  getFilesystemSize,
  getFilesystemUsed,
  getMaxFileDescriptors,
  getAllocatedFileDescriptors
}

File: discovery/src/monitors/filesystem.py
import psutil

# We first check '/var/k8s' in case the service operator has elected to
# mount an external volume for k8s data. Otherwise, default to the root path at '/'


def get_filesystem_size(**kwargs):
    """
    Gets the size of the entire filesystem (bytes)

    Kwargs:
        db: global database instance
        redis: global redis instance
    """
    disk = None
    try:
        disk = psutil.disk_usage("/var/k8s")
    except:
        disk = psutil.disk_usage("/")

    return disk.total


def get_filesystem_used(**kwargs):
    """
    Gets the used portion of the filesystem (bytes)

    Kwargs:
        db: global database instance
        redis: global redis instance
    """
    disk = None
    try:
        disk = psutil.disk_usage("/var/k8s")
    except:
        disk = psutil.disk_usage("/")

    return disk.used
